livre/core/pipeline/AsyncData.cpp

```diff
@@ -0,0 +1,171 @@
+
+namespace
+{
+    /* Boost wait_for_any implementation uses the internal locks from the futures.
+       Therefore, the get/set/query operations on promises/futures causes deadlocks
+       when used with wait_for_any( futurelist ) if futurelist is including the
+       future to be queried. With below implementation no locking is needed between
+       future operations. A higher granularity can be added for checking whether
+       the list includes the future or not, but simply exiting the wait_for_any
+       operation when lock is owned is simpler.
+     */
+    ReadWriteMutex waitForAnyLock;
+}
+
+struct AsyncData::Impl
+{
+    Impl( const DataInfo& dataInfo )
+        : _future( _promise.get_future( ))
+        , _dataInfo( dataInfo )
+    {}
+
+    ~Impl()
+    {
+        if( !isReady())
+            _promise.set_value( PortDataPtr( ));
+    }
+
+    const PortDataPtr& get( const std::type_index& dataType ) const
+    {
+        ReadLock lock( waitForAnyLock );
+
+        const PortDataPtr& data = _future.get();
+
+        if( !data )
+            LBTHROW( std::runtime_error( "Returns empty data" ));
+
+        if( data->dataType != dataType )
+            LBTHROW( std::runtime_error( "Types does not match on get value"));
+
+        return data;
+    }
+
+    void set( const PortDataPtr& data )
+    {
+        if( data )
+        {
+            if( _dataInfo.second != data->dataType )
+                LBTHROW( std::runtime_error( "Types does not match on set value"));
+        }
+
+        ReadLock lock( waitForAnyLock );
+        if( !_future.is_ready( ))
+            _promise.set_value( data );
+    }
+
+    bool isReady() const
+    {
+        ReadLock lock( waitForAnyLock );
+        return _future.is_ready();
+    }
+
+    void wait() const
+    {
+        ReadLock lock( waitForAnyLock );
+        _future.wait();
+    }
+
+    void reset()
+    {
+        ConstPortDataPromise newPromise;
+        _promise.swap( newPromise );
+        _future = _promise.get_future();
+    }
+
+    ConstPortDataPromise _promise;
+    mutable ConstPortDataFuture _future;
+    const DataInfo _dataInfo;
+};
+
+AsyncData::AsyncData( const DataInfo& dataInfo )
+    :  _impl( new Impl( dataInfo ))
+{}
+
+const PortDataPtr& AsyncData::get( const std::type_index& dataType ) const
+{
+    return _impl->get( dataType );
+}
+
+AsyncData::~AsyncData()
+{}
+
+const std::type_index& AsyncData::getDataType() const
+{
+    return _impl->_dataInfo.second;
+}
+
+const std::string& AsyncData::getName() const
+{
+    return _impl->_dataInfo.first;
+}
+
+void AsyncData::set( const PortDataPtr& data )
+{
+    _impl->set( data );
+}
+
+bool AsyncData::isReady() const
+{
+    return _impl->isReady();
+}
+
+void AsyncData::wait() const
+{
+    _impl->wait();
+}
+
+void AsyncData::reset()
+{
+    _impl->reset();
+}
+
+bool waitForAny( const Futures& futures )
+{
+    if( futures.empty( ))
+        return false;
```

Comment:
I did it because you can use in a loop for instance. while( waitforany( ...  )).
But I have changed the implmentation a bit and now it seems a bit obsolote !
You are right